src/osd/OSD.cc

```diff
@@ -6301,17 +6301,22 @@ int OSD::probe_smart_device(const char *device, int timeout, std::string *result
   ret = output.read_fd(smartctl.get_stdout(), 100*1024);
   if (ret < 0) {
     derr << "failed read from smartctl: " << cpp_strerror(-ret) << dendl;
-    return ret;
+  } else if (ret == 0) {
```

Comment:
no need to handle this here. it's been taken care of in `OSD::probe_smart()`.

Reply:
@tchaikov ret==0 case needs handling. this is exactly thre root cause of the  issue reported. please see the backtrace in tracker.

Reply:
okay, probably we should put something like
```c++
if (ret < 0) {
  //...
  return ret;
}
*result = output.to_str();
dout(10) << "smartctl output is: " << *result << dendl;
```

Reply:
@tchaikov not following your point.  still calling smartctl.join() in multiple places?

Reply:
@guzhongyan this comment has nothing to do with the join() call, i am trying to remove the `ret == 0` check.

Reply:
@tchaikov following your comments, then my understanding is:
if (ret <= 0) {
  //... call smartctl.join()
  return ret;
}

Am i right?

Reply:
sorry, i was wrong: should not `return ret` at that moment. i'd suggest:

```diff
diff --git a/src/osd/OSD.cc b/src/osd/OSD.cc
index 16f99ba582..e87623a0c6 100644
--- a/src/osd/OSD.cc
+++ b/src/osd/OSD.cc
@@ -6301,12 +6301,9 @@ int OSD::probe_smart_device(const char *device, int timeout, std::string *result
   ret = output.read_fd(smartctl.get_stdout(), 100*1024);
   if (ret < 0) {
     derr << "failed read from smartctl: " << cpp_strerror(-ret) << dendl;
-  } else if (ret == 0) {
-    derr << "failed read from smartctl: zero size"  << dendl;
-    ret = -EINVAL;
   } else {
-    dout(10) << "smartctl output is: " << output.c_str() << dendl;
-    *result = output.c_str();
+    *result = output.to_str();
+    dout(10) << "smartctl output is: " << *result << dendl;
   }

   if (smartctl.join() != 0) {
```

Reply:
@tchaikov  updated. please review.